Source/WebGPU/WebGPU/Buffer.mm

```diff
@@ -111,10 +111,38 @@ static MTLStorageMode storageMode(bool deviceHasUnifiedMemory, WGPUBufferUsageFl
     return MTLStorageModePrivate;
 }
 
+static NSUInteger heapSizeForResource(MTLSizeAndAlign sizeAndAlign)
+{
+    ASSERT(sizeAndAlign.size > 0);
+    constexpr NSUInteger objectsPerHeap = 32;
+    constexpr NSUInteger limitedHeapSize = 4 * MB;
+    auto alignedSize = roundUpToMultipleOfNonPowerOfTwo(sizeAndAlign.align, sizeAndAlign.size);
+    return std::max(alignedSize, std::min(alignedSize * objectsPerHeap, limitedHeapSize));
+}
+
 id<MTLBuffer> Device::safeCreateBuffer(NSUInteger length, MTLStorageMode storageMode, MTLCPUCacheMode cpuCacheMode, MTLHazardTrackingMode hazardTrackingMode) const
 {
     MTLResourceOptions resourceOptions = (cpuCacheMode << MTLResourceCPUCacheModeShift) | (storageMode << MTLResourceStorageModeShift) | (hazardTrackingMode << MTLResourceHazardTrackingModeShift);
-    id<MTLBuffer> buffer = [m_device newBufferWithLength:std::max<NSUInteger>(1, length) options:resourceOptions];
+
+    length = std::max<NSUInteger>(1, length);
+    MTLSizeAndAlign sizeAndAlign = [m_device heapBufferSizeAndAlignWithLength:length options:resourceOptions];
+    id<MTLHeap> heap;
+    for (id<MTLHeap> h in [m_heaps reverseObjectEnumerator]) {
+        if (h.storageMode == storageMode && sizeAndAlign.size <= [h maxAvailableSizeWithAlignment:sizeAndAlign.align]) {
+            heap = h;
+            break;
+        }
+    }
```

Comment:
I think this would read better as a separate helper function. Then the break could be a return.

Reply:
It may be counterintuitive, but it's better to iterate in forward order. In the simple case, all your heaps are full except the last one, and the iteration is wasted work. But once you start freeing things, iterating in forward order tends to compact memory over time, while iterating in reverse order tends to fragment memory over time.